### PR TITLE
Various error reporting fixes plus Maintenance Current fix.

### PR DESCRIPTION
**Pull request by neunhoef**
Mostly: Make sure that an ex-leader who still believes to be a leader
does not report strange stuff to Current. Added preconditions and
checks.

### CODE REVIEW

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/5046/

File: arangod/Cluster/FollowerInfo.cpp [hunk 1]
```diff
@@ -279,7 +283,6 @@ bool FollowerInfo::remove(ServerID const& sid) {
   } while (TRI_microtime() < startTime + 30 &&
            application_features::ApplicationServer::isRetryOK());
   if (!success) {
-    _followers = _oldFollowers;
```
Comment (maierlars) [inline arangod/Cluster/FollowerInfo.cpp]: What about checking Plan before updating Current?

File: arangod/Cluster/FollowerInfo.cpp [hunk 2]
```diff
@@ -191,38 +194,39 @@ bool FollowerInfo::remove(ServerID const& sid) {
...
+      for (auto const& i : *_followers) {
+        if (i != sid) {
+          v->push_back(i);
+        }
       }
     }
-  }
-  auto _oldFollowers = _followers;
-  _followers = v;  // will cast to std::vector<ServerID> const
+    auto _oldFollowers = _followers;
```
Comment (maierlars) [inline arangod/Cluster/FollowerInfo.cpp]: should be unused now

File: arangod/Cluster/Maintenance.cpp [hunk 3]
```diff
@@ -525,6 +525,7 @@ arangodb::Result arangodb::maintenance::executePlan(VPackSlice const& plan,
   std::vector<ActionDescription> actions;
   report.add(VPackValue(AGENCY));
   {
+    // TODO: Just putting an empty array does not make any sense here!
```
Comment (maierlars) [inline arangod/Cluster/Maintenance.cpp]: ?

File: arangod/Cluster/Maintenance.cpp [hunk 4]
```diff
@@ -838,6 +840,29 @@ arangodb::Result arangodb::maintenance::reportInCurrent(
...
+            << "Ooops, we have a shard for which we believe to be the leader,"
+               " but the Plan does not have it any more, we do not report in "
+               "Current about this, database: " << dbName
+            << ", shard: " << shName;
+          continue;
+        }
+
+        VPackSlice thePlanList = pdbs.get(planPath);
+        if (!thePlanList.isArray() || thePlanList.length() == 0 ||
+            thePlanList[0].copyString() != serverId) {
```
Comment (maierlars) [inline arangod/Cluster/Maintenance.cpp]: There is `isEqualString`. We can eliminate most of the copies.

File: arangod/Cluster/Maintenance.cpp [hunk 5]
```diff
@@ -852,31 +877,21 @@ arangodb::Result arangodb::maintenance::reportInCurrent(
...
-              if (!oldValue.isNone()) {
-                report.add(VPackValue("precondition"));
-                {
-                  VPackObjectBuilder p(&report);
-                  report.add(
-                    PLAN_COLLECTIONS + dbName + "/" + colName + "/shards/" + shName,
-                    oldValue);
-                }
-              }
-            } catch(...) {
```
Comment (maierlars) [inline arangod/Cluster/Maintenance.cpp]: This try ... catch is now no longer needed?

File: arangod/Cluster/Maintenance.cpp (hunk 3 above)
Comment (neunhoef) [reply arangod/Cluster/Maintenance.cpp]: ! Left for later cleanup.

File: arangod/Cluster/FollowerInfo.cpp [hunk 6]
```diff
@@ -140,29 +147,41 @@ void FollowerInfo::add(ServerID const& sid) {
...
-            AgencyOperation(key, AgencyValueOperationType::SET, newValue.slice()));
-        trx.operations.push_back(
-            AgencyOperation("Current/Version", AgencySimpleOperationType::INCREMENT_OP));
-        AgencyCommResult res2 = ac.sendTransactionWithFailover(trx);
-        if (res2.successful()) {
-          success = true;
-          break;  //
+        if (!planEntry.isArray() || planEntry.length() == 0 ||
+            !planEntry[0].isString() ||
+            planEntry[0].compareString(ServerState::instance()->getId()) != 0) {
```
Comment (maierlars) [inline arangod/Cluster/FollowerInfo.cpp]: `equalString?

**Comment by maierlars** [comment]
LGTM

File: arangod/Cluster/Maintenance.cpp (hunk 5 above)
Comment (neunhoef) [reply arangod/Cluster/Maintenance.cpp]: Indeed, because we have checked the plan beforehand.

File: arangod/Cluster/FollowerInfo.cpp (hunk 1 above)
Comment (neunhoef) [reply arangod/Cluster/FollowerInfo.cpp]: Does not really matter, we hold the lock, and do the check before communications and roll back if the plan is bad.

**Pull request merged**